linera-storage/src/lib.rs

```diff
@@ -273,14 +279,63 @@ pub trait Store: Sized {
     }
 }
 
+#[cfg(any(feature = "wasmer", feature = "wasmtime"))]
+async fn read_publish_bytecode_operation(
+    store: &impl Store,
+    application_description: &UserApplicationDescription,
+) -> Result<SystemOperation, ExecutionError> {
+    let UserApplicationDescription {
+        bytecode_id,
+        bytecode_location,
+        ..
+    } = application_description;
+    let value = store
+        .read_value(bytecode_location.certificate_hash)
+        .await
+        .map_err(|error| match error {
+            ViewError::NotFound(_) => ExecutionError::ApplicationBytecodeNotFound(Box::new(
+                application_description.clone(),
+            )),
+            _ => error.into(),
+        })?
+        .into_inner();
+    let mut operations = match value {
+        CertificateValue::ConfirmedBlock { executed_block, .. } => executed_block.block.operations,
+        _ => return Err(ExecutionError::InvalidBytecodeId(*bytecode_id)),
+    };
+    let index = usize::try_from(bytecode_location.operation_index)
+        .map_err(|_| linera_base::data_types::ArithmeticError::Overflow)?;
+    linera_base::ensure!(
+        index < operations.len(),
+        ExecutionError::InvalidBytecodeId(*bytecode_id)
+    );
+    match operations.swap_remove(index) {
```

Comment:
(This would avoid swapping the last element into the `index`-th position. Not sure if it's actually better.)

```suggestion
    match operations.into_iter().nth(index).unwrap() {
```

Then you could omit the `unwrap()` and match on `Some`, so the `ensure!` above isn't needed anymore.

Reply:
`into_iter().nth()` is not O(1)

Reply:
I think it is: [The default implementation](https://github.com/rust-lang/rust/blob/46ecc10c6951a2a0e52d93fe5d3acae9743e3ab9/library/core/src/iter/traits/iterator.rs#L396) uses `advance_by`, and [`Vec`'s `advance_by` looks like it's O(1)](https://github.com/rust-lang/rust/blob/46ecc10c6951a2a0e52d93fe5d3acae9743e3ab9/library/alloc/src/vec/into_iter.rs#L217).

Reply:
@afck I do see an efficient `advance_by` here: https://doc.rust-lang.org/src/alloc/vec/into_iter.rs.html
So you're probably right that it's O(1)